java/src/com/facebook/watchman/WatchmanClientImpl.java

```diff
@@ -255,11 +255,13 @@ public void call(Map<String, Object> message) throws Exception {
             .name(subscriptionId)
             .root(root)
             .build();
-        if (!subscriptions.containsKey(subscription)) {
+
+        Callback listener = subscriptions.get(subscription);
+        if (listener == null) {
```

Comment:
I agree that you've correctly identified that it's possible to have a null value for a key in the subscriptions map.

But it looks to me like the only way that's possible is if a user called the `subscribe()` method with a null listener.  (Is there any legit reason to do that?  That seems like a usage error to me.)  What path did you observe that made an NPE possible?  Looking at `unsubscribe()`, I don't see a path that would cause an NPE here – it looks to me that it's correctly waiting for cancellation confirmation from watchman before removing the entry.

I don't necessarily mind this change here, since it's perhaps slightly more efficient to save a double-lookup of the key in the map (once for `containsKey` and once for `get`), but how about we fix the real problem at the source and add `Preconditions.checkNotNull(listener)` after line 160, so that it's an invariant that the subscriptions map doesn't contain any null values?